src/services/httpparsing.cpp

```diff
@@ -96,67 +50,20 @@ static std::string getPostData(std::istream& in, int content_length) {
 	return query;
 }
 
-static void parseKeyValuePair(const std::string& q, std::map<std::string, std::string>& kvp, bool unescape = false, const std::string& delim = "=") {
-	std::string::size_type sep = q.find_first_of(delim);
-
-	std::string key, val;
-
-	if (sep == std::string::npos) {
-		key = q;
-		val = ""; // Empty value
-	} else {
-		key = urldecode(q.substr(0, sep));
-		val = urldecode(q.substr(sep + 1, q.length() - (sep + 1)));
-		trim(val);
-	}
-
-	trim(key);
-	if (key.empty())
-		return;
-	std::transform(key.begin(), key.end(), key.begin(), ::tolower);
-
-	if (unescape && val.size() >= 2) {
-
-		if (val.front() == '\"' && val.back() == '\"') {
-			val.erase(0, 1);
-			val.erase(val.length() - 1);
-		}
-	}
-
-	kvp[key] = val;
-}
-
 /**
  * Parses a query string into a given Params structure.
  */
 void parseQuery(const std::string& query, Parameters &params) {
 	if (query.length() == 0)
 		return;
 
-	// see RFC 3986 ch. 3.4
-
-	/*
-	 // query string is already extracted, no need for this part.
-	 auto beg = q.find_first_of("?");
-	 if(beg == std::string::npos) // No query string present
-	 return;
-	 auto end = q.find_first_of("#", beg);
-	 if(end == std::string::npos) // No fragment
-	 end = q.length() - 1;
-	 */
-
-	std::string::size_type last = 0;
-	std::string::size_type cur;
-
-	while ((cur = query.find_first_of("&", last)) != std::string::npos) {
-		std::string sub = query.substr(last, cur - last);
-		parseKeyValuePair(sub, params);
-		last = cur + 1;
-	}
-
-	if (last < query.length()) {
-		std::string sub = query.substr(last, query.length() - last);
-		parseKeyValuePair(sub, params);
+	Poco::URI uri;
+    //setRawQuery instead of setQuery because setQuery will encode the given string, but here it is already encoded.
+	uri.setRawQuery(query);
+	auto queryParameters = uri.getQueryParameters();
```

Comment:
why do we still need our own Parameter class? Can't we use the Poco `QueryParameters` directly?

Reply:
[QueryParameters](https://pocoproject.org/docs/Poco.URI.html#18968) is a vector: `typedef std::vector < std::pair < std::string, std::string > > QueryParameters;`. So we would lose the map functionality.

Reply:
You mean the multimap functionality? Ok, we need it because we want to aggregate multiple values for a key. So the getAll methods for Parameter are fine.
However, I don't like the getXYZ and getYAZ*Last* methods. I think we agreed to always take the *last* value for a key when there are multiple entries (override semantics). This behavior should be uniform for all service, thus there should be no need for getting both first and last.
If you agree, please remove the getLast methods and make the get-method to return the last value for the given key.

Also for the future if you add new public methods to a class please attach a small doc comment explaining what the function does. I know lots of existing methods miss this, but we should try to avoid this now ;)

Reply:
Yes, you're right, I will adjust the get-method and will add comments.